### ISSUE TITLE
Imported electricity doesn't fill the gap between total production and total demand

### ISSUE THREAD

**Issue by redekok** [quintel/etsource#1944]
In an electrification [scenario](https://pro.energytransitionmodel.com/scenarios/404433) for Muide Meulestede (Gent), we expected that whenever the total electricity production doesn't meet the total electricity demand, imported electricity would be used to fill this gap. This didn't seem to be the case though. As can be seen in the screenshot below, we observed that at least 9 kW of electricity can be imported per hour (see black circle). However, at the spot where the red circle is drawn, the total electricity demand is 7 kW per hour but only 2 kW of electricity is imported.

@antw Would you have any idea why this is the case?

![screenshot 2019-03-04 at 10 45 00](https://user-images.githubusercontent.com/5946498/53726110-94c7f300-3e6d-11e9-933b-743209d29594.png)

**Comment by antw** [comment]
Hmm, this is disheartening. 😕

The cause is an optimisation to reduce the amount of data we have to send to the browser when displaying curves (the reduction was by as much as 75%).

The reason is each value in each curve is truncated to its integer value. So 12716.12 becomes 12716, 33221.9 becomes 33221, and so on. Previously we were sending many decimal places which had no impact on the chart, but significantly slowed down response times.

The trick works well *except* in smaller datasets where the loads are in the single-digit MW range; a value like 3.9 gets truncated to 3 and the errors mean the chart no longer makes sense.

The best I can do without throwing away the optimisation – or coming up with some special-case for small datasets – is to *round* to the nearest whole number instead of truncating. The result is better, but still not perfect:

![screen shot 2019-03-04 at 16 06 56](https://user-images.githubusercontent.com/4383/53745822-b342e400-3e97-11e9-947b-270d796ff585.png)

Including @ChaelKruip

**Comment by ChaelKruip** [comment]
> The best I can do without throwing away the optimisation – or coming up with some special-case for small datasets – is to round to the nearest whole number instead of truncating. 

Could we make use of the fact that we have units in the ETM that adapt to the nearest factor of 1000?

In the case above, it seems that the result is rounded when expressed in **MW**: 3.9-> 3. Later, the result is displayed as 3000 **kW**. Could we round the number in kW (which might be 3000.56) to 3000 and send that integer to the browser instead?

**Comment by antw** [comment]
> Could we round the number in kW (which might be 3000.56) to 3000 and send that integer to the browser instead?

It would require some benchmarking to figure out whether the cost is worth it. We'd have to update all the load curve queries to multiply each value in the curve by 1000; in my synthetic benchmarks, multiplying a curve and truncating is about 2.3x slower than just truncating. Then we've got an extra three significant figures to send for every hour; that's an extra 26kb (uncompressed) *per curve*.

I'm happy to experiment with this to see how well it performs, but I expect my preference will be to disable the optimisation entirely for smaller datasets; it seems like a shame to lose this improvement for larger datasets which are the majority of scenarios.

**Comment by ChaelKruip** [comment]
> We'd have to update all the load curve queries to multiply each value in the curve by 1000

Sorry, I think I was not clear in describing my suggestion. Multiplying all curves seems costly indeed and I would not propose to go down that road.

Let me try to make myself more clear 😄. I imagine the ETM can work in one of two ways:
1. ETEngine/Fever/Merit calculates curves in MW, casts to integer, and sends this result to the browser. The browser converts MW to kW for small datasets.
2. ETEngine/Fever/Merit calculates curves in MW and converts MW to kW for small datasets, casts this number to integer, and sends this result to the browser.

@antw am I correct to assume/guess that the first option is (roughly) what we currently do?
I think that because for case (1), rounding the number in MW results in loss of precision for the result expressed in kW.
Would it be possible to do the second thing? In that case sending integer values to the browser would not result in (noticeable) loss of precision right?

**Comment by antw** [comment]
> ETEngine/Fever/Merit calculates curves in MW and converts MW to kW for small datasets, casts this number to integer, and sends this result to the browser.

Correct me if I'm wrong, but  I think you're suggesting to special-case small datasets to provide extra precision? That is what I briefly suggested in the first post (but didn't describe in any detail). We can do this:

1. We need to define what is considered to be a small region (e.g. < 5000 residences?).
2. Instead of truncating to an integer we can round to a couple of decimal places. This is similarly 2.5x slower than simply truncating, but I expect it's an acceptable trade-off for small datasets.

The result is much improved:

![screen shot 2019-03-06 at 15 31 58](https://user-images.githubusercontent.com/4383/53892913-18781000-4025-11e9-8b6f-74ecd258f18f.png)

**Comment by michieldenhaan** [comment]
> We need to define what is considered to be a small region (e.g. < 5000 residences?).

I think the issue occurs for bigger datasets as well. The image below is about hydrogen (not electricity) for Capelle aan den IJssel, which has about 30k residences (with 100% H2 HHPs):

![pro energytransitionmodel com - energy transition model - your free independent comprehensive fact-based scenario builder 2019-03-06 16-46-30](https://user-images.githubusercontent.com/32056448/53893937-7ad60e00-402f-11e9-82c3-08b744581aaf.png)

If a user wants to build a more moderate scenario (e.g. a 2030 scenario in which only limited H2 is available) the issue can also arise for a city like Amsterdam (700k residences). 15% H2 HHPs in Amsterdam:

![pro energytransitionmodel com - energy transition model - your free independent comprehensive fact-based scenario builder 2019-03-06 16-51-11](https://user-images.githubusercontent.com/32056448/53894249-12d3f780-4030-11e9-9ed2-397e78f38831.png)

**Comment by ChaelKruip** [comment]
> Correct me if I'm wrong, but I think you're suggesting to special-case small datasets to provide extra precision? That is what I briefly suggested in the first post (but didn't describe in any detail).

Not really. I'm trying (and failing 😆 ) to suggest that we use the unit-scaling functionality (W, kW, MW, GW etc.) to scale the result **prior to rounding it to integer values** and sending it to the browser. Does that make sense?

**Comment by redekok** [comment]
What's the status of this issue? Is it worth investigating other scaling vs. rounding "methods" in order to solve this issue? Is this something we want to fix prior to the deploy? @antw @ChaelKruip

**Comment by ChaelKruip** [comment]
> Is this something we want to fix prior to the deploy? 

It's not critical for the deploy IMO. Not a bad idea to pick this up soon after though as far as I'm concerned.

**Comment by MartLubben** [comment]
Is this the same issue. This case seems a bit more extreme than I see above. I have opened Elburg. 

![Screenshot 2019-07-10 at 13 34 22](https://user-images.githubusercontent.com/33415156/60966026-9b80de00-a317-11e9-8a92-67b47d36a86c.png)

**Comment by michieldenhaan** [comment]
That migt be a bug. Total demand minus the four series shown in the graph results in a residual demand:

![beta-engine energytransitionmodel com - EnergyTransitionModel 2019-07-10 13-55-25](https://user-images.githubusercontent.com/32056448/60967380-ce78a100-a31a-11e9-89e3-c9bc8ed143e4.png)

**Comment by antw** [comment]
> Is this the same issue.

It could be the same, in my opinion. If a lot of the series have demands < 0.5 MW, the query will truncate that to zero. It's quite possible that the residual demand is caused by this truncation.

**Comment by michieldenhaan** [comment]
Even when adding all series (not just the ones displayed) it doesn't sum up to the total 🤔. It might be some dataset quirk, it seems to happen for a lot of ``om_`` datasets but not for other local datasets.. The ``om_`` datasets are not exported from ETLocal but are made with Excel.

``Q(merit_households_other_demand).sum + Q(merit_buildings_other_demand).sum  + Q(merit_ev_demand).sum + Q(merit_households_hot_water_demand).sum  + Q(merit_households_space_heating_demand).sum + Q(merit_industry_metals_demand).sum + Q(merit_industry_chemical_demand).sum + Q(merit_buildings_space_heating_demand).sum + Q(merit_agriculture_demand).sum + Q(merit_industry_other_demand).sum + Q(merit_transport_other_demand).sum + Q(merit_other_demand).sum  + Q(merit_hv_network_loss + Q(merit_other_flexibility_demand).sum``
